woocommerce/woocommerce
# Fix - Express payments that have errors cause the whole block to stop working

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:





Closes #42256.

With this PR, we're enhancing


### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Install and activate [WooCommerce PayPal Payments](https://wordpress.org/plugins/woocommerce-paypal-payments/) - set it up and ensure it's working on your site, ensure the express payment button renders in the editor and the front end. 
2. Clone the development version of [WooCommerce Stripe Payment Gateway](https://github.com/woocommerce/woocommerce-gateway-stripe) 
3. Build the plugin, following the instructions in the repo, then activate and enable it. 
4. Run `npm run start` in the stripe plugin directory.
5. Ensure Stripe's express payment button renders in the editor and the front end.
6. Open the plugin in your editor, specifically at this file: https://github.com/woocommerce/woocommerce-gateway-stripe/blob/8ffd22aff3b06eda02a1ae2fd8368b71450b36a9/client/blocks/payment-request/index.js#L11
7. Redefine `ApplePayPreview` like so:
```ts
const ApplePayPreview = () => {
const x = window.null.something.broken;
return <img src={ applePayImage } alt="" disabled={x} />;
}
```
8. Go to https://github.com/woocommerce/woocommerce-gateway-stripe/blob/4f47751f5dd3d26eb245878a936f96510a11d2e4/client/blocks/payment-request/payment-request-express.js#L95C1-L95C1 and add:
```ts
  const y = window.null.broken;
  console.log( y );
```
9. Open the site editor, go to Templates > Checkout
10. Confirm the error notice and PayPal button is visible in the express checkout block. 
11. Open the site editor and go to Templates > Cart.
9. Confirm the error notice and PayPal button is visible in the express checkout block. 
10. Visit the site as admin.
12. Add a product to the Cart and visit the Cart Block page.
13. Confirm the error notice and PayPal button is visible in the express checkout block. 
14. Navigate to the Checkout block page.
15. Confirm the error notice and PayPal button is visible in the express checkout block. 
16. Visit your site as a Guest. 
17. Add a product to the Cart and visit the Cart Block page.
18. Confirm only the PayPal button is visible in the express checkout block. 
19. Navigate to the Checkout block page.
20. Confirm only the PayPal button is visible in the express checkout block. 


https://github.com/woocommerce/woocommerce/assets/11503784/f0f7c83a-8f85-44c8-ba51-5a0ee033163c




### Changelog entry




<details>

#### Significance




#### Type




#### Message 
Improve handling of express checkout: Now, error messages for specific express payment methods won't affect the functionality of others.

#### Comment 

</details>
